### ISSUE TITLE
[bug] .multinomial() for cuda.DoubleTensor

### ISSUE THREAD

**Issue by donglixp** [pytorch/pytorch#2044]
The following code can reproduce the bug. The line "torch.multinomial(a.exp(), 1)" could work properly. But after casting into DoubleTensor, the sampled indexes will be out of range.

```python
import torch

a = torch.rand(20000).float().mul(-20).view(20,1000).clone().cuda()
print(a)

# this returns correct results
print(torch.multinomial(a.exp(), 1))
# incorrect!
print(torch.multinomial(a.double().exp_(), 1))
```

```
-15.8601 -14.0862 -11.3351  ...  -11.1707 -10.3008  -0.7116
-18.1648  -9.0676  -0.4030  ...  -15.8194 -10.5331  -9.2996
-12.7967  -5.2371 -18.6317  ...   -0.1992 -15.9085 -15.9299
           ...               ⋱              ...
 -9.2137  -2.4529  -2.9928  ...  -14.3643 -12.9492 -17.5611
 -5.5278  -2.8773 -19.5927  ...  -11.5547  -4.2149 -18.7635
 -3.0814  -6.3477  -0.3227  ...   -7.2171 -15.2311  -0.4129
[torch.cuda.FloatTensor of size 20x1000 (GPU 0)]


  723
  261
  657
  663
  851
  764
  245
  417
  405
  210
  420
  479
  213
  919
  532
  742
  994
  416
  176
  833
[torch.cuda.LongTensor of size 20x1 (GPU 0)]


 3.8407e+18
 3.5479e+18
 4.0363e+18
 3.7260e+18
 4.1986e+18
 4.2638e+18
 4.5605e+18
 4.4725e+18
 3.5909e+18
 3.8631e+18
 4.5026e+18
 4.2295e+18
 4.5749e+18
 4.3938e+18
 4.2139e+18
 4.3064e+18
 4.3244e+18
 3.9574e+18
 4.0812e+18
 4.5084e+18
[torch.cuda.LongTensor of size 20x1 (GPU 0)]
```

**Comment by pavanky** [comment]
@donglixp does this happen without CUDA ? If not may be edit the title to be more specific.

**Comment by donglixp** [comment]
@pavanky The code snippet works well on CPU. I have edited the title according to the suggestion. Thanks!

**Comment by amartya18x** [comment]
@pavanky I tried it without cuda. It didn't happen with me.

**Comment by soumith** [comment]
@donglixp what pytorch version are you using? `print(torch.__version__)`

**Comment by donglixp** [comment]
@soumith 0.1.12_2 w/ CUDA 8.0, python 2.7

**Comment by soumith** [comment]
Thanks, I can reproduce this on master. Can look and figure it out.

**Comment by soumith** [comment]
cc: @killeent who is most familiar with multinomial code.

**Comment by killeent** [comment]
Odd, as we have fallback code that handles an error where we don't select a multinomial, so something strange must be happening here.

**Comment by BestSonny** [comment]
#2072 Seems it is related to limited shared memory.

**Comment by soumith** [comment]
fixed via #2072 thanks to @BestSonny !

**Comment by donglixp** [comment]
Thanks for the rapid fix!
